File: go4java/Composition.java
import java.util.Collection;
import java.util.ArrayList;

class Composition {
    class Task {
        private String message;

        public Task(String message) {
            this.message = message;
        }

        public void Run() {
            System.out.println("running " + this.message);
        }
    }

    class Runner {
        private String name;

        public Runner(String name) {
            this.name = name;
        }

        public String getName() {
            return this.name;
        }

        public void Run(Task task) {
            task.Run();
        }

        public void RunAll(Task[] tasks) {
            for (Task task : tasks) {
                Run(task);
            }
        }
    }

    // START_COUNTING OMIT
    class CountingRunner {
        private Runner runner;
        private int count;

        public CountingRunner(String message) {
            this.runner = new Runner(message);
            this.count = 0;
        }

        public void Run(Task task) {
            count++;
            runner.Run(task);
        }

        public void RunAll(Task[] tasks) {
            count += tasks.length;
            runner.RunAll(tasks);
        }

        public int getCount() { return count; }

        public String getName() { return runner.getName(); }
    }
    // END_COUNTING OMIT

    public void test() {
        // START_MAIN OMIT
        CountingRunner runner = new CountingRunner("my runner");

        Task[] tasks = { new Task("one"), new Task("two"), new Task("three")};
        runner.RunAll(tasks);

        System.out.printf("%s ran %d tasks\n", runner.getName(), runner.getCount());
        // END_MAIN OMIT
    }

    public static void main(String[] args) {
        new Composition().test();
    }
}

File: go4java/Composition.go
package main

import "fmt"

// TASK OMIT
type Task struct{ msg string }

func (t Task) Run() { fmt.Println("running", t.msg) }

// RUNNER OMIT
type Runner struct{ name string }

func NewRunner(name string) *Runner { return &Runner{name} }
func (r *Runner) Name() string      { return r.name }
func (r *Runner) Run(t Task)        { t.Run() }

func (r *Runner) RunAll(ts []Task) {
	for _, t := range ts {
		r.Run(t)
	}
}

// COUNTING OMIT
type CountingRunner struct {
	runner *Runner // HL
	count  int
}

func NewCountingRunner(name string) *CountingRunner {
	return &CountingRunner{runner: NewRunner(name)}
}

func (r *CountingRunner) Run(t Task) {
	r.count++
	r.runner.Run(t)
}

func (r *CountingRunner) RunAll(ts []Task) {
	r.count += len(ts)
	r.runner.RunAll(ts)
}

func (r *CountingRunner) Count() int { return r.count }

func (r *CountingRunner) Name() string { return r.runner.Name() }

// MAIN OMIT
func main() {
	r := NewCountingRunner("my runner")
	tasks := []Task{{"one"}, {"two"}, {"three"}}
	r.RunAll(tasks)
	fmt.Printf("%s ran %d tasks\n", r.Name(), r.Count())
}
